### PR TITLE
Features/initial impl new api

### PR DESCRIPTION
**Pull request by nahguam**
# stream-registry PR

Initial implementation of core features along with Domain, Schema & Stream entities.

# PR Checklist Forms

- [x] CHANGELOG.md updated
- [x] Reviewer assigned
- [x] PR assigned (presumably to submitter)
- [x] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/predicate/VersionPredicateFactory.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import com.expediagroup.streamplatform.streamregistry.model.Stream;
+
+@Component
+public class VersionPredicateFactory {
+  public java.util.stream.Stream<Stream> filter(Stream query, java.util.stream.Stream<Stream> stream) {
+    if (query.getVersion() == null) {
+      return stream;
+    } else if (query.getVersion() < 0) {
+      throw new IllegalArgumentException("version must be greate than or equal to 0");
```
Comment (nahguam) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/predicate/VersionPredicateFactory.java]: "Version must be greater than or equal to 0."

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/service/StreamService.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  private final NameDomainPatternMatchPredicateFactory nameDomainPatternMatchPredicateFactory;
+  private final VersionPredicateFactory versionPredicateFactory;
+
+  @Override
+  public void upsert(Stream stream) {
+    Optional<Stream> existing = streamRepository.get(stream.key());
+    entityValidator.validate(stream, existing);
+    configuredEntityValidator.validate(stream, existing);
+    domainConfiguredEntityValidator.validate(stream, existing);
+
```
Comment (nahguam) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/service/StreamService.java]: remove extra whitespace line

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/service/StreamService.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    Optional
+        .ofNullable(stream.getVersion())
+        .filter(v -> v > 0)
+        .orElseThrow(() -> new IllegalArgumentException("Version must be provided and greater than 0."));
+    existing.ifPresent(e -> checkArgument(
+        stream.getVersion() > e.getVersion(),
+        "Version greater than the current version."));
```
Comment (nahguam) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/service/StreamService.java]: Double check whether the `Stream.Key` includes version, and hence if the logic here is valid.

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/validator/EntityValidator.java [hunk 4]
```diff
@@ -13,13 +13,24 @@
...
 
+@Component
+@RequiredArgsConstructor
+public class EntityValidator {
+  private final NameValidator nameValidator;
+
+  public void validate(Entity<?> entity, Optional<? extends Entity<?>> existing) {
+    nameValidator.validate(entity.getName());
+
+    //TODo owner is mandatory and exists
```
Comment (nahguam) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/validator/EntityValidator.java]: TOD**O**

File: graphql-api/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/resolver/Query.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  public List<GraphQLSchema> schemas(
+      String name,
+      String owner,
+      String description,
+      Iterable<GraphQLKeyValue> tags,
+      GraphQLConfiguration configuration,
+      String domain) {
+    System.out.println(Schema
```
Comment (nahguam) [inline graphql-api/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/resolver/Query.java]: remove sysout

File: graphql-api/src/main/resources/application.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+graphql:
```
Comment (nahguam) [inline graphql-api/src/main/resources/application.yaml]: Could this yaml file live in the `graph-api` module?

File: graphql-client/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/client/ReactorApollo.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+import com.apollographql.apollo.ApolloCall;
+import com.apollographql.apollo.ApolloCall.StatusEvent;
+import com.apollographql.apollo.ApolloSubscriptionCall;
+import com.apollographql.apollo.api.Response;
+import com.apollographql.apollo.exception.ApolloException;
+import com.apollographql.apollo.internal.subscription.ApolloSubscriptionTerminatedException;
+import com.apollographql.apollo.internal.util.Cancelable;
+
+public final class ReactorApollo {
```
Comment (nahguam) [inline graphql-client/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/client/ReactorApollo.java]: This should live in a separate module - even better, contributed to apollographql as `apollo-reactor-support`

File: graphql-client/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/client/ReactorApollo.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+  public static <T> Flux<Response<T>> from(ApolloSubscriptionCall<T> call, OverflowStrategy overflowStrategy) {
+    return Flux.create(sink -> {
+      sink.onCancel(cancel(call));
+      call.execute(new FluxSinkCallback(sink));
+    }, overflowStrategy);
+  }
+
+  private static Disposable cancel(Cancelable cancelable) {
+    return () -> cancelable.cancel();
```
Comment (nahguam) [inline graphql-client/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/client/ReactorApollo.java]: `cancelable::cancel`?

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/egsp/EgspConfluentSchemaHandler.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.springframework.stereotype.Component;
+
+import com.expediagroup.streamplatform.streamregistry.handler.Handler;
+import com.expediagroup.streamplatform.streamregistry.model.Configuration;
+import com.expediagroup.streamplatform.streamregistry.model.Schema;
+
+@Component
+public class EgspConfluentSchemaHandler implements Handler<Schema> {
+  public static final String TYPE = "egsp.confluent";
```
Comment (nahguam) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/egsp/EgspConfluentSchemaHandler.java]: `.alpha`

File: graphql-client/pom.xml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  <properties>
+    <introspectionFile>${project.build.directory}/graphql-schema/schema.json</introspectionFile>
+    <clientPackage>com.expediagroup.streamplatform.streamregistry.graphql.client</clientPackage>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.apollographql.apollo</groupId>
+      <artifactId>apollo-runtime</artifactId>
+      <version>1.0.0</version>
```
Comment (nahguam) [inline graphql-client/pom.xml]: I think we need to add http://jcenter.bintray.com repo as `1.0.0` is not available on central

File: graphql-api/src/main/resources/application.yaml (hunk 6 above)
Comment (nahguam) [reply graphql-api/src/main/resources/application.yaml]: it does already...

**Comment by prabhakar1983** [comment]
`mvn clean install` throws me the following error
```
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-maven) @ stream-registry ---
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireMavenVersion failed with message:
Detected Maven Version: 3.5.0 is not in the allowed range 3.6.0.
```

File: graphql-api/src/main/resources/stream-registry.graphql [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  upsertStream(
+    name: String!
+    description: String!
+    tags: [KeyValueInput!]!
+    configuration: ConfigurationInput!
+    domain: String!
+    version: Int!
+    schema: NameDomainInput!
+  ): Boolean!
+}
```
Comment (neoword) [inline graphql-api/src/main/resources/stream-registry.graphql]: NOICE!
😄 👏 🎉 🍾 🍻

File: graphql-api/src/main/resources/stream-registry.graphql [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  streams(
+    name: String
+    owner: String
+    description: String
+    tags: [KeyValueInput!]
+    configuration: ConfigurationInput
+    domain: String
+    version: Int
+    schema: NameDomainInput
+  ): [Stream!]!
```
Comment (neoword) [inline graphql-api/src/main/resources/stream-registry.graphql]: This is a great first pass.

For future releases, lets create an issue to Support Pagination. Need to be able to support cardinality of greater > 100.
https://graphql.org/learn/pagination/

**Pull request merged**